source/common/ssl/tls_certificate_config_impl.cc

```diff
@@ -0,0 +1,19 @@
+#include "common/ssl/tls_certificate_config_impl.h"
+
+#include "common/config/datasource.h"
+
+namespace Envoy {
+namespace Ssl {
+
+TlsCertificateConfigImpl::TlsCertificateConfigImpl(const envoy::api::v2::auth::Secret& config)
+    : message_([&config] {
+        envoy::api::v2::auth::Secret message;
+        message.CopyFrom(config);
+        return message;
+      }()),
+      name_(config.name()), certificate_chain_(Config::DataSource::read(
+                                config.tls_certificate().certificate_chain(), true)),
```

Comment:
The way you have structured the code, there is no guarantee that `tls_certificate()` is actually defined. You need more error checking here.

I don't feel that strongly about it, but I would in general prefer we do away w/ the dynamic casting and just have a base secret interface, and then sub-class interfaces for each secret type, and then potentially a way to lookup a secret in the manager by type. I think that would consolidate a lot of the error checking logic.

Reply:
Created multi layered map and added template functions that add and lookup secret based on the type.
map<key, map<name, secret>>